zipline/src/jsTest/kotlin/app/cash/zipline/DynamicFunctionTest.kt

```diff
@@ -0,0 +1,274 @@
+    val dynamicEcho = echo.asDynamicSuspendingFunction()
+
+    val deferred: Deferred<Any?> = async {
+      val request = js("""{message:"suspend this call!"}""")
+      return@async dynamicEcho(client, listOf(request))
+    }
+
+    assertThat(requests.receive()).isEqualTo(EchoRequest("suspend this call!"))
+    responses.send(EchoResponse("yo, suspended caller"))
+    assertThat(JSON.stringify(deferred.await()))
+      .isEqualTo("""{"message":"yo, suspended caller"}""")
+  }
+
+  /** This test exercises the synchronous return case for suspending calls. */
+  @Test
+  fun suspendedFunctionNotSuspended() = runTest {
+    val (endpointA, endpointB) = newEndpointPair(this)
+
+    val service = object : SuspendingEchoService {
+      override suspend fun suspendingEcho(request: EchoRequest): EchoResponse {
+        assertThat(request).isEqualTo(EchoRequest("don't suspend this call!"))
+        return EchoResponse("yo, unsuspended caller")
+      }
+    }
+
+    endpointA.bind<SuspendingEchoService>("service", service)
+    val client = endpointB.take<SuspendingEchoService>("service")
+
+    val echo = client.sourceType!!.functions.first { "suspendingEcho" in it.signature }
+    val dynamicEcho = echo.asDynamicSuspendingFunction()
+
+    val request = js("""{message:"don't suspend this call!"}""")
+    val response = dynamicEcho(client, listOf(request))
+
+    assertThat(JSON.stringify(response))
+      .isEqualTo("""{"message":"yo, unsuspended caller"}""")
+  }
+
+  @Test
+  fun unstructuredValues() = runTest {
+    acceptAndProduce(
+      kotlinValue = "hello",
+      jsValue = "hello",
+    )
+    acceptAndProduce<Int>(
+      kotlinValue = 5,
+      jsValue = 5,
+    )
+    acceptAndProduce<Double>(
+      kotlinValue = 2147483648.0,
+      jsValue = 2147483648.0,
+    )
+    acceptAndProduce<Long>(
+      kotlinValue = 2147483648L,
+      jsValue = js("2147483648"),
+    )
+    acceptAndProduce<Long>(
+      kotlinValue = 9007199254740991L,
+      jsValue = js("9007199254740991"),
+    )
+    acceptAndProduce<Boolean>(
+      kotlinValue = true,
+      jsValue = true,
+    )
+    acceptAndProduce<String?>(
+      kotlinValue = null,
+      jsValue = null,
+    )
+  }
+
+  @Test
+  fun collections() = runTest {
+    acceptAndProduce<List<String>>(
+      kotlinValue = listOf(element = "hello"),
+      jsValue = js("""["hello"]"""),
+    )
+    acceptAndProduce<List<String?>>(
+      kotlinValue = listOf("hello", null),
+      jsValue = js("""["hello", null]"""),
+    )
+  }
+
+  @Test
+  fun objects() = runTest {
+    acceptAndProduce<EchoRequest>(
+      kotlinValue = EchoRequest(message = "hello"),
+      jsValue = js("""{"message":"hello"}"""),
+    )
+    acceptAndProduce<EchoRequest?>(
+      kotlinValue = null,
+      jsValue = null,
+    )
+  }
+
+  /** Large longs can't be serialized due to a potential precision loss. */
+  @Test @Ignore
+  fun encodeDecodeLargeLong() = runTest {
```

Comment:
You can work around this with a contextual serializer, like this one:
https://github.com/cashapp/zipline/blob/23ec7fb557a9314a9d0c1d7a703f5dcaf9e7ca11/zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/longs.kt#L32